targets/TARGET_STM/i2c_api.c

```diff
@@ -831,36 +973,45 @@ int i2c_write(i2c_t *obj, int address, const char *data, int length, int stop)
 
     i2c_ev_err_enable(obj, i2c_get_irq_handler(obj));
 
-    ret = HAL_I2C_Master_Sequential_Transmit_IT(handle, address, (uint8_t *) data, length, obj_s->XferOperation);
+    const HAL_StatusTypeDef status = HAL_I2C_Master_Sequential_Transmit_IT(
+        handle, address, (uint8_t *)data, length, obj_s->XferOperation);
 
-    if (ret == HAL_OK) {
-        timeout = BYTE_TIMEOUT_US * (length + 1);
-        /*  transfer started : wait completion or timeout */
-        while (!(obj_s->event & I2C_EVENT_ALL) && (--timeout != 0)) {
-            wait_us(1);
-        }
+    if (status != HAL_OK) {
+      DEBUG_PRINTF("ERROR in i2c_write\r\n");
 
-        i2c_ev_err_disable(obj);
+      return I2C_ERROR_BUS_BUSY;
+    }
 
-        if ((timeout == 0) || (obj_s->event != I2C_EVENT_TRANSFER_COMPLETE)) {
-            DEBUG_PRINTF(" TIMEOUT or error in i2c_write\r\n");
-            /* re-init IP to try and get back in a working state */
-            i2c_init(obj, obj_s->sda, obj_s->scl);
-        } else {
-            count = length;
-        }
-    } else {
-        DEBUG_PRINTF("ERROR in i2c_read\r\n");
+    uint32_t timeout = (BYTE_TIMEOUT_US * (length + 1));
+
+    /*  transfer started : wait completion or timeout */
+    while (!(obj_s->event & I2C_EVENT_ALL) && (--timeout != 0)) {
+        wait_us(1);
+    }
+
+    i2c_ev_err_disable(obj);
+
+    if ((timeout == 0) || (obj_s->event != I2C_EVENT_TRANSFER_COMPLETE)) {
```

Comment:
@scartmell-arm 
I think that inside this `if` we should detect error type and return appropriately